network/p2p/capabilities.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p/core/discovery"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libpeerstore "github.com/libp2p/go-libp2p/core/peerstore"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	algoDht "github.com/algorand/go-algorand/network/p2p/dht"
+	"github.com/algorand/go-algorand/network/p2p/peerstore"
+)
+
+// Capability represents functions that some nodes may provide and other nodes would want to know about
+type Capability string
+
+const (
+	// Archival nodes
+	Archival Capability = "archival"
+	// Catchpoints storing nodes
+	Catchpoints = "catchpointStoring"
+)
+
+const operationTimeout = time.Second * 5
+const advertisementInterval = time.Hour * 22
+
+// CapabilitiesDiscovery exposes Discovery interfaces and wraps underlying DHT methods to provide capabilities advertisement for the node
+type CapabilitiesDiscovery struct {
+	disc discovery.Discovery
+	dht  *dht.IpfsDHT
+	log  logging.Logger
+	wg   sync.WaitGroup
+}
+
+// Advertise implements the discovery.Discovery/discovery.Advertiser interface
+func (c *CapabilitiesDiscovery) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	return c.disc.Advertise(ctx, ns, opts...)
+}
+
+// FindPeers implements the discovery.Discovery/discovery.Discoverer interface
+func (c *CapabilitiesDiscovery) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	return c.disc.FindPeers(ctx, ns, opts...)
+}
+
+// Close should be called when fully shutting down the node
+func (c *CapabilitiesDiscovery) Close() {
+	_ = c.dht.Close()
+	c.wg.Wait()
+}
+
+// Host exposes the underlying libp2p host.Host object
+func (c *CapabilitiesDiscovery) Host() host.Host {
+	return c.dht.Host()
+}
+
+// AddPeer adds a given peer.AddrInfo to the Host's Peerstore, and the DHT's routing table
+func (c *CapabilitiesDiscovery) AddPeer(p peer.AddrInfo) (bool, error) {
+	c.Host().Peerstore().AddAddrs(p.ID, p.Addrs, libpeerstore.TempAddrTTL)
```

Comment:
nit: maybe use libp2p.AddressTTL to be consistent with what's used in peerstore initialization?